### PR TITLE
ingress syncer agent

### PR DESCRIPTION
**Pull request by philbrookes**
agent running on workload clusters, watches ingress and annotates DNS records in the control plane with required changes.

verification:
```
make local-setup MCTC_WORKLOAD_CLUSTERS_COUNT=1
```

Update the `config/manager/manager.yaml` to contain the necessary environment variable and update the image policy (N.B. Do not copy/paste this, the hyphens from github are bad for the YAML parser):
```
    containers:
        - command:
            - /controller
            args:
            - --leader-elect
            image: controller:latest
            imagePullPolicy: Never
            env:
            - name: AWS_ACCESS_KEY_ID
            value: <AWS_ACCESS_KEY_ID>
            - name: AWS_SECRET_ACCESS_KEY
            value: <AWS_SECRET_ACCESS_KEY>
            - name: AWS_DNS_PUBLIC_ZONE_ID
            value: <AWS_DNS_PUBLIC_ZONE_ID>
            - name: ZONE_ROOT_DOMAIN
            value: <ZONE_ROOT_DOMAIN>
``` 
Load the image into kind:
```
export KUBECONFIG=./hack/kubeconfigs/mctc-control-plane.kubeconfig
make kind-load-controller
```
deploy the image:
```
make deploy-controller
```
Target the workload cluster you wish to run on:
```sh
export KUBECONFIG=./hack/kubeconfigs/mctc-workload-1.kubeconfig
```
Create the control plane secret
Update the `samples/syncer/secret.yaml` from the `hack/kubeconfigs/mctc-control-plane.kubeconfig` and apply it:
```sh
kubectl create -f samples/syncer/secret.yaml
```
Run the agent locally:
```sh
make run-agent
```
Create the sample application (in a second terminal also targetting the workload cluster):
```sh
export KUBECONFIG=./hack/kubeconfigs/mctc-workload-1.kubeconfig
kubectl apply -f ./samples/syncer/application.yaml
```
Wait for  the agent to log:
```
certificate secret in place for  host adding dns endpoints	{"host": "s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net"}
```
Then after a few moments (for propagation) dig the host:
```
dig s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net
```
once it is resolving, curl it:
```
curl -k https://s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net
```

### CODE REVIEW

**Comment by david-martin** [comment]
Some cmds I've used to mount env vars from configmap & secret instead of modifying the deployment

```
kubectl -n multi-cluster-traffic-controller-system patch deployment mctc-controller-manager -p '{"spec": {"template": {"spec":{"containers":[{"name":"manager","imagePullPolicy":"Never", "envFrom": [{"configMapRef":{"name":"controller-config"}}, {"secretRef":{"name":"aws-credentials"}}] }]}}}}' -n multi-cluster-traffic-controller-system

kubectl -n multi-cluster-traffic-controller-system create configmap controller-config $(for envvar in $(cat controller-config.env); do echo -n "--from-literal=$envvar ";done)

kubectl -n multi-cluster-traffic-controller-system create secret generic aws-credentials $(for envvar in $(cat aws-credentials.env); do echo -n "--from-literal=$envvar ";done)
```

**Review by maleck13 — APPROVED** [review]
mostly minor things. Verified via steps. As the goal is to move toward a more generic sync ok with this as an intermediary step to remove the need for the controller

File: cmd/agent/main.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	//+kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+const (
+	defaultCtrlNS       = "argocd"
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply cmd/agent/main.go:43]: yeah, this seems like a good idea.

File: pkg/controllers/ingress/ingress.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+		return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, fmt.Errorf("error creating control plane client: %v", err.Error())
+	}
+	current := &networkingv1.Ingress{}
+	err = r.Client.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, current)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	target := current.DeepCopy()
+	accessor := traffic.NewIngress(target)
+	res, err := r.Handle(ctx, accessor)
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply pkg/controllers/ingress/ingress.go:92]: The Reconcile function was getting quite large, and pulling out the modifications to the CR to a separate function made it far easier to reason about for me.

File: pkg/controllers/ingress/ingress.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		targets = append(targets, target.Value)
+	}
+
+	//adding DNS patches
+	if len(targets) > 0 && trafficAccessor.GetDeletionTimestamp() == nil {
+		err = r.DNS.PatchTargets(ctx, targets, trafficAccessor.GetHosts(), r.ClusterID, false)
+		if err != nil {
+			return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, err
+		}
+		controllerutil.AddFinalizer(trafficAccessor, trafficFinalizer)
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply pkg/controllers/ingress/ingress.go:168]: done.

File: config/rbac/role.yaml [hunk 4]
```diff
@@ -133,13 +121,3 @@ rules:
   - get
   - patch
   - update
-- apiGroups:
```
Comment (david-martin) [inline config/rbac/role.yaml:136]: Is there a separate set of roles for the agent that includes ingress?
Can kubebuilder be used for the agent too?

**Review by david-martin — CHANGES_REQUESTED** [review]
Functionality wise, it seems to be doing what's intended.
The code looks reasonable.

1 concern I have with the split is how much it brings the project away from being able to use kubebuilder.
Is the intention to still have all the various kubebuilder cmds and generation work for the main controller?
Do we want a way for the agent to work with kubebuilder too?
I noticed for example the change in name from `manager` to `controller` in 1 place,
but the kustomize setup still uses `manager.yaml`.
Also, we're missing various yaml files for the agent that would usually be scaffolded by kubebuilder.
I think we should figure out how we're going to deploy the agent i.e. where do the kustomize files go, and how will they be maintained separately from the main controller.
I'm wondering if having the agent in this repo is right move given then concerns above.

File: config/rbac/role.yaml (hunk 4 above)
Comment (philbrookes) [reply config/rbac/role.yaml:136]: good point. I'll make and test some now, and add it to this PR (or it could be a separate PR to keep this one moving?)

Comment (david-martin) [reply config/rbac/role.yaml:136]: For now, let's capture an issue for the deploy of the agent.
Priority wise, deploying it to a cluster is lower than moving on to the syncer piece

**Review by david-martin — APPROVED** [review]
Approving after further reflection and some offline discussion.
Issues to be captured for future work:

- deploy of the agent to a cluster
- target for loading the image into a cluster, upding the pull policy to never
- mounting env vars from a secret & configmap

**Pull request merged**